### DISCUSSION TITLE
Can array builders be provided at runtime/gc level?

### DISCUSSION THREAD

**Discussion by En3Tho** [dotnet/runtime#46914] [Ideas] [answered]
Methods like Enumerable.ToArray or structures like ImmutableArrayBuilder, StringBuilder internally use array resizing.
Could this resizing logic actually be implemented in GC in form of some RuntimeArrayBuilder? It could support methods like Add(T) or AddMany(ReadOnlySpan<T>), or AddMany(Seq<T>) and so on. In array resizing would actually be managed by GC in some kind of pool/list system. The idea is that GC could potentially avoid copying as it might know whether there is still memory left in a chunk/bucket. And if after building there is memory left in the chunk, GC could reuse that memory block later. Or is possible to make or it's just way too complex and instead could bring too much trouble with fragmentation management?
@Maoni0 @jkotas

**Comment by jkotas** [comment, marked as answer]
This was discussed multiple times before e.g. https://github.com/dotnet/runtime/issues/38263#issuecomment-648159073

Unlike C/C++ allocators, .NET GC packs memory blocks next to each other. Chance that there is an adjacent free memory block available is very low.

**Reply by En3Tho** [reply]
Thanks. Should have searched better.
